README
trytond_songbook
================

The songbook module of the Tryton application platform.

This module creates a music database for use by DJ and Karaoke service
business.

Installing
----------

See INSTALL

Support
-------

If you encounter any problems with Tryton, please don't hesitate to ask
questions on the Tryton bug tracker, mailing list, wiki or IRC channel:

  http://bugs.tryton.org/
  http://groups.tryton.org/
  http://wiki.tryton.org/
  irc://irc.freenode.net/tryton

License
-------

See LICENSE

Copyright
---------

See COPYRIGHT


For more information please visit the Tryton web site:

  http://www.tryton.org/

For more information about this module in particular:

  mailto:info@coalesco.ca



CHANGELOG
Version 3.0.0 - 2014-06-20
* Initial release
Version 3.0.1 - 2014-11-21
* Schema additions for "tracks"
* Improved UI


song.py
from trytond.model import ModelView, ModelSQL, fields
from trytond.pool import Pool
from nereid import (
    request, abort, render_template, login_required, url_for, flash, jsonify,
    current_app, route
)
from nereid.contrib.pagination import Pagination, BasePagination
from nereid.ctx import has_request_context

__all__ = ['Song']


class Song(ModelSQL, ModelView):
    "Song"
    __name__ = "songbook.song"

    title = fields.Char('Title', 64, required=True, select=True)
    artist = fields.Many2One("songbook.artist", 'Artist', required=True)
    tracks = fields.One2Many(
        "songbook.track",
        'song',
        'Tracks of This Song'
    )

    def get_rec_name(self, name):
        return " | ".join(
            x for x in [self.title, self.artist.full_name] if x
        )

    def serialize(self):
        """
        Serialize the song object and return a dictionary.
        """
        object_json = {
            "url": url_for(
                'songbook.song.render_html',
                id=self.id,
            ),
            "objectType": self.__name__,
            "id": self.id,
            "title": self.title,
            "artist": self.artist.full_name,
        }
        return object_json

    @classmethod
    def __setup__(cls):
        super(Song, cls).__setup__()
        cls._sql_constraints = [
            ('name_artist_uniq', 'UNIQUE(title, artist)',
             'This song by this artist already exists in the system.')
        ]
        cls._order.insert(0, ('title', 'ASC'))

    @classmethod
    @route('/songbook/api/songs', methods=['GET', 'POST'])
    def call_api_index(cls):
        """
        JSON-formatted REST API to support 3rd party integration, apps
        and web page javascript such as search-as-you-type.
        """
        artist_filter = '%' + request.args.get('artistcontains', '') + '%'
        title_filter = request.args.get('titlestartswith', '') \
            + '%' + request.args.get('titlecontains', '') + '%'
        domain = [
            ('title', 'ilike', title_filter),
            ('artist.full_name', 'ilike', artist_filter)
        ]
        songs = cls.search(domain, limit=int(request.args.get('limit', '5')))
        return jsonify(
            songs=[s.serialize() for s in songs]
        )

    @classmethod
    @route('/songbook/songs/<int:id>', methods=['GET'])
    def render_html(cls, id=0):
        """
        output details of a selected song to web client
        """
        song=cls.browse([id])[0]
        return render_template(
            'songbook_song-detail.jinja',
            song=song
        )

    @classmethod
    @route('/songbook/songs', methods=['GET', 'POST'])
    def render_html_index(cls):
        """
        output song list to web client
        """
        artist_filter = '%' + request.args.get('artistcontains', '') + '%'
        title_filter = request.args.get('titlestartswith', '') \
            + '%' + request.args.get('titlecontains', '') + '%'
        page = request.args.get('page', 1, int)
        domain = [
            ('title', 'ilike', title_filter),
            ('artist.full_name', 'ilike', artist_filter)
        ]
        songs = Pagination(
            cls, domain, page, 25
        )

        return render_template(
            'songbook_song-list.jinja',
            songs=songs
        )



album.py
from trytond.model import ModelView, ModelSQL, fields
from trytond.pool import Pool
from nereid import (
    request, abort, render_template, login_required, url_for, flash, jsonify,
    current_app, route
)

__all__ = ['Album']


class Album(ModelSQL, ModelView):
    "Album"
    __name__ = "songbook.album"

    songbook = fields.Many2One("songbook.songbook", 'Songbook', required=True)
    code = fields.Char('Code', 16, required=True, select=True)
    name = fields.Char('Name', 64, required=True, select=True)
    publisher = fields.Many2One("songbook.publisher", 'Publisher', required=True)
    description = fields.Text('Description')
    tracks = fields.One2Many(
        "songbook.track",
        'album',
        'Tracks in This Album'
    )

    @classmethod
    def __setup__(cls):
        super(Album, cls).__setup__()
        cls._sql_constraints = [
            ('songbook_code_uniq', 'UNIQUE(songbook, code)',
             'This album code already exists in the specified songbook.')
        ]
        cls._order.insert(0, ('songbook', 'ASC'))
        cls._order.insert(1, ('code', 'ASC'))

    @classmethod
    @route('/songbook/albums/<code>.txt', methods=['GET'])
    def call_api_index(cls, code):
        """
        Delimited text file via website
        """
        album=cls.search([('code', '=', code)])[0]
        return render_template(
            'songbook_songlist-txt.jinja',
            tracks = album.tracks
        )



track.py
from trytond.model import ModelView, ModelSQL, fields
from trytond.pool import Pool
from nereid import (
    request, abort, render_template, login_required, url_for, flash, jsonify,
    current_app, route
)
from nereid.contrib.pagination import Pagination, BasePagination
from nereid.ctx import has_request_context

__all__ = ['Track']


class Track(ModelSQL, ModelView):
    "Track"
    __name__ = "songbook.track"

    album = fields.Many2One("songbook.album", 'Album', required=True)
    code = fields.Char('Code', 16, required=True, select=True)
    song = fields.Many2One("songbook.song", 'Song', required=True)

    @classmethod
    def __setup__(cls):
        super(Track, cls).__setup__()
        cls._sql_constraints = [
            ('album_code_uniq', 'UNIQUE(album, code)',
             'This track code already exists in the specified album.')
        ]
        cls._order.insert(0, ('code', 'ASC'))




artist.py
from trytond.model import ModelView, ModelSQL, fields
from trytond.pool import Pool

from sql import Table, Column, Literal, Desc, Asc, Expression, Flavor
from sql.functions import Now, Extract
from sql.operators import Or, And, Concat, ILike, Operator
from sql.conditionals import Coalesce
from sql.aggregate import Count, Max

from nereid import (
    request, abort, render_template, login_required, url_for, flash, jsonify,
    current_app, route
)
from nereid.contrib.pagination import Pagination, BasePagination
from nereid.ctx import has_request_context

__all__ = ['Artist']


class Artist(ModelSQL, ModelView):
    "Artist"
    __name__ = "songbook.artist"
    _rec_name = 'full_name'

    last_name = fields.Char('Last Name', required=True, select=True)
    first_name = fields.Char('First Name', select=True)
    full_name = fields.Function(
        fields.Char('Full Name'), 'get_full_name',
        searcher='search_full_name'
    )
    rev_name = fields.Function(
        fields.Char('Reversed Name'), 'get_rev_name'
    )
    songs = fields.One2Many(
        "songbook.song",
        'artist',
        'Songs by This Artist'
    )

    def serialize(self):
        """
        Serialize the artist object and return a dictionary.
        """
        object_json = {
            "url": url_for(
                'songbook.artist.render_html',
                id=self.id,
            ),
            "objectType": self.__name__,
            "id": self.id,
            "lastName": self.last_name,
            "firstName": self.first_name,
            "fullName": self.full_name,
        }
        return object_json

    @classmethod
    def __setup__(cls):
        super(Artist, cls).__setup__()
        cls._sql_constraints = [
            ('name_uniq', 'UNIQUE(last_name, first_name)',
             'An artist with that name already exists.')
        ]
        cls._order.insert(0, ('last_name', 'ASC'))
        cls._order.insert(1, ('first_name', 'ASC'))

    @classmethod
    def search_full_name(cls, name, clause):
        "Search Full Name"
        _, operator, value = clause
        Operator = fields.SQL_OPERATORS[operator]
        table = cls.__table__()
        fullname = Concat(
            Coalesce(table.first_name, Literal('')),
            Concat(
                Literal(' '),
                Coalesce(table.last_name, Literal(''))
            )
        )

        query = table.select(table.id, where=Operator(fullname, value))
        return [('id', 'in', query)] 

    def get_full_name(self, name):
        if self.first_name is None:
            fullname = self.last_name
        else:
            fullname = "%s %s" % (self.first_name, self.last_name) 
        return fullname.strip(" ")

    def get_rev_name(self, name):
        if self.first_name is None:
            revname = self.last_name
        else:
            revname = "%s, %s" % (self.last_name, self.first_name) 
        return revname.strip(", ")

    @classmethod
    @route('/songbook/api/artists', methods=['GET', 'POST'])
    def call_api_index(cls):
        """
        JSON-formatted REST API to support 3rd party integration, apps
        and web page javascript such as search-as-you-type.
        """
        name_filter = '%' + request.args.get('namecontains', '') + '%'
        domain = [
            ('full_name', 'ilike', name_filter)
        ]
        artists = cls.search(domain, limit=int(request.args.get('limit', '5')))
        return jsonify(
            artists=[a.serialize() for a in artists]
        )

    @classmethod
    @route('/songbook/artists/<int:id>', methods=['GET'])
    def render_html(cls, id=0):
        """
        output details of a selected artist to web client
        """
        artist=cls.browse([id])[0]
        return render_template(
            'songbook_artist-detail.jinja',
            artist=artist
        )

    @classmethod
    @route('/songbook/artists', methods=['GET', 'POST'])
    def render_html_index(cls):
        """
        output artist list to web client
        """
        name_filter = '%' + request.args.get('namecontains', '') + '%'
        page = request.args.get('page', 1, int)
        domain = [
            ('full_name', 'ilike', name_filter)
        ]
        artists = Pagination(
            cls, domain, page, 25
        )

        return render_template(
            'songbook_artist-list.jinja',
            artists=artists
        )



songbook.py
from trytond.model import ModelView, ModelSQL, fields
from trytond.report import Report
from trytond.transaction import Transaction
from trytond.pool import Pool
from trytond.wizard import Wizard, StateView, Button, StateTransition
from sql import Table, As, Literal, Column, Desc, Asc, Expression, Flavor
from sql.conditionals import Coalesce
from sql.operators import Or, And, Concat, ILike, In, Operator
from nereid import (
    request, abort, render_template, login_required, url_for, flash, jsonify,
    current_app, route
)
from nereid.contrib.pagination import Pagination, BasePagination

__all__ = [
    'Songbook',
    'ExportTracks',
    'ExportTracksStart',
    'ExportTracksResult',
    'SongbookByArtist'
]


class Songbook(ModelSQL, ModelView):
    "Songbook"
    __name__ = "songbook.songbook"

    name = fields.Char('Name', 64, required=True, select=True)
    description = fields.Text('Description')
    albums = fields.One2Many(
        'songbook.album',
        'songbook',
        'Albums in This Songbook'
    )
    songs_title = fields.Function(
        fields.One2Many(
            'songbook.song',
            None,
            'Songs by Title in This Songbook'
        ), 'get_songs_title'
    )
    songs_artist = fields.Function(
        fields.One2Many(
            'songbook.song',
            None,
            'Songs by Artist in This Songbook'
        ), 'get_songs_artist'
    )

    @classmethod
    def __setup__(cls):
        super(Songbook, cls).__setup__()
        cls._sql_constraints = [
            ('name_uniq', 'UNIQUE(name)',
             'A songbook with that name already exists.')
        ]
        cls._order.insert(0, ('name', 'ASC'))

    def get_songs_title(self, name):
        songdict = {}
        songlist = []

        for album in self.albums:
            for track in album.tracks:
                if track.song.rec_name not in songdict:
                    songdict[track.song.rec_name] = track.song.id

        for song in sorted(songdict):
            songlist.append(songdict[song])

        return songlist

    def get_songs_artist(self, name):
        artdict = {}
        artlist = []

        for album in self.albums:
            for track in album.tracks:
                songkey = track.song.artist.rev_name + ' | ' + track.song.title
                if songkey not in artdict:
                    artdict[songkey] = track.song.id

        for artist in sorted(artdict):
            artlist.append(artdict[artist])

        return artlist

    @classmethod
    @route('/songbook/songbooks/<int:id>')
    def render_html(cls, id=1):
        """
        output songbook home page to client
        """
        songbook=cls.browse([id])[0]
        return render_template(
            'songbook_songbook-detail.jinja',
            songbook=songbook
        )

    @classmethod
    @route('/songbook/songbooks', methods=['GET', 'POST'])
    def render_html_index(cls):
        """
        output song list to web client
        """
        name_filter = '%' + request.args.get('namecontains', '') + '%'
        page = request.args.get('page', 1, int)
        domain = [
            ('name', 'ilike', name_filter)
        ]
        songbooks = Pagination(
            cls, domain, page, 25
        )

        return render_template(
            'songbook_songbook-list.jinja',
            songbooks=songbooks
        )

    @classmethod
    @route('/songbook')
    def render_html_home(cls):
        """
        output the home page of the songbook web app to client
        """

        return render_template('songbook_home.jinja')

class ExportTracks(Wizard):
    "Export Tracks in Songbook"
    __name__ = 'songbook.songbook.export_tracks'

    start = StateView(
        'songbook.songbook.export_tracks.start',
        'songbook.songbook_export_tracks_start_view_form', [
            Button('Cancel', 'end', 'tryton-cancel'),
            Button('Export', 'export', 'tryton-ok', default=True),
        ]
    )
    export = StateTransition()
    result = StateView(
        'songbook.songbook.export_tracks.result',
        'songbook.songbook_export_tracks_result_view_form', [
            Button('Close', 'end', 'tryton-cancel'),
        ]
    )

    def transition_export(self):
        """
        Delimited text file for import into CAVS or similar jukebox
        """
        pool = Pool()
        Song = pool.get('songbook.song')
        Artist = pool.get('songbook.artist')
        Publisher = pool.get('songbook.publisher')
        Album = pool.get('songbook.album')
        Track = pool.get('songbook.track')

        cursor = Transaction().cursor

        song = Song.__table__()
        artist = Artist.__table__()
        publisher = Publisher.__table__()
        album = Album.__table__()
        track = Track.__table__()

        songbook_ids = Transaction().context.get('active_ids')

        title_and_publisher = Concat(
            song.title, Concat(
                Literal(' .'), Concat(publisher.code, Literal('.'))
            )
        )

        artist_fullname = Concat(
            Coalesce(artist.first_name, Literal('')),
            Concat(
                Literal(' '),
                Coalesce(artist.last_name, Literal(''))
            )
        )

        export_select = track.join(
            album, condition=(album.id == track.album)
        ).join(
            publisher, condition=(publisher.id == album.publisher)
        ).join(
            song, condition=(song.id == track.song)
        ).join(
            artist, condition=(artist.id == song.artist)
        ).select(
            track.code, title_and_publisher, artist_fullname,
            where=In(album.songbook, songbook_ids)
        )
        export_select.order_by = Asc(track.code)

        cursor.execute(*export_select)

        rv_list = ['|'.join(x) for x in cursor.fetchall()]

        self.result.file = buffer(bytearray(u'\r\n'.join(rv_list), 'utf-8'))
        return 'result'

    def default_result(self, fields):
        file_ = self.result.file
        self.result.file = False  # No need to store it in session
        return {
            'file': file_,
        }

class ExportTracksStart(ModelView):
    "Export Tracks in Songbook"
    __name__ = 'songbook.songbook.export_tracks.start'

class ExportTracksResult(ModelView):
    "Export Tracks in Songbook"
    __name__ = 'songbook.songbook.export_tracks.result'

    file = fields.Binary('File', readonly=True)

class SongbookByArtist(Report):
    __name__ = 'songbook.songs_by_artist'

    @classmethod
    def parse(cls, report, objects, data, localcontext):
        artlist = dict((s.id, []) for s in objects)
        for songbook in objects:
            artdict = {}
            for album in songbook.albums:
                for track in album.tracks:
                    if track.song.artist.rev_name not in artdict:
                        artdict[track.song.artist.rev_name] = {}
                    if track.song.title not in artdict[
                        track.song.artist.rev_name
                    ]:
                        artdict[track.song.artist.rev_name][
                            track.song.title
                        ] = track.song.id

            for artkey in sorted(artdict):
                artlist[songbook.id].append({
                    'rev_name': artkey,
                    'songs': sorted(artdict[artkey]),
                })

        localcontext['artists'] = lambda songbook_id: artlist[songbook_id]
        res = super(SongbookByArtist, cls).parse(
            report, objects, data, localcontext
        )
        return res

